### ISSUE TITLE
Crash after internet disconnection

### ISSUE THREAD

**Issue by Rayope** [TwosHusbandS/Project-127#132]
Sometimes, the launcher (Project-1.27) crashes when internet randomly disconnects.
The last time, I had the game running.
Logs:

> === Project - 127 Started (Version: '1.2.2.3' BuildInfo: '1.2.2.3 - Build 1' Built at: '2021-03-07T09:20:36' Central European Time) ===

...

> [2021-05-17T11:36:18] - Settings initiated and loaded. Will stop Logging

> [2021-05-17T11:36:18] - GetStringFromURL failed. Probably Network related. URL = 'https://raw.githubusercontent.com/TwosHusbandS/Project-127/master/Installer/Update.xml'

> [2021-05-17T11:36:18] - - e.ToString():
System.Net.Http.HttpRequestException: An error occurred while sending the request. ---> System.Net.WebException: The remote name could not be resolved: 'raw.githubusercontent.com'
   at System.Net.HttpWebRequest.EndGetResponse(IAsyncResult asyncResult)
   at System.Net.Http.HttpClientHandler.GetResponseCallback(IAsyncResult ar)
   --- End of inner exception stack trace ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter1.GetResult()
   at Project_127.HelperClasses.FileHandling.GetStringFromURL(String pURL, Boolean surpressPopup)

> [2021-05-17T11:36:18] - - e.Message.ToString():
An error occurred while sending the request.

> [2021-05-17T11:37:04] - GetStringFromURL failed. Probably Network related. URL = 'https://raw.githubusercontent.com/TwosHusbandS/Project-127/master/Installer/Update.xml'

> [2021-05-17T11:37:04] - - e.ToString():
System.Net.Http.HttpRequestException: An error occurred while sending the request. ---> System.Net.WebException: The remote name could not be resolved: 'raw.githubusercontent.com'
   at System.Net.HttpWebRequest.EndGetResponse(IAsyncResult asyncResult)
   at System.Net.Http.HttpClientHandler.GetResponseCallback(IAsyncResult ar)
   --- End of inner exception stack trace ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter1.GetResult()
   at Project_127.HelperClasses.FileHandling.GetStringFromURL(String pURL, Boolean surpressPopup)

> [2021-05-17T11:37:04] - - e.Message.ToString():
An error occurred while sending the request.`

**Comment by TwosHusbandS** [comment]
Does that happen on specific Pages or Clicks / Button Presses? Or just while not doing anything inside Project 1.27?

Also, does the crashes line up with the time in the log messages? Or could the logs be from something else (few seconds or minutes before the crash)?.

Do you have a way of reproducing the crash?

Currently looking into it.

**Comment by Rayope** [comment]
Answered somewhere else but for reference:
It happens while I'm not doing anything inside Project 1.27, it's just opened in the background.
For reproducing... I just did this (and Project 1.27 didn't log anything this time, maybe because I didn't click 'Debug' from the crash window)
* I connect my computer to internet
* I launch Project 1.27
* Project 1.27 starts the social club, and I sign in.
* I launch windowed GTA V through Project 1.27
* I tab out and do other stuff, not touching neither GTAV nor Project 1.27 (both running in the background)
* I intentionnally disconnect internet
* ... 28min30sec later
* Project 1.27 crash with the following window:
![ScreenShot025](https://user-images.githubusercontent.com/84181546/119137163-ca525a00-ba40-11eb-85c7-a2739643c32b.jpg)
![ScreenShot026](https://user-images.githubusercontent.com/84181546/119137175-cd4d4a80-ba40-11eb-9342-a3f7e56b950d.jpg)

Last time it logged something (maybe because I clicked 'Debug'), it was at the time of start...

**Comment by Rayope** [comment]
Well, an even better and more consistent way to reproduce, just in case:
* Start Project 1.27
* Connect to Social Club
* Disconnect internet
* Click "Launch GTA V"
--> Project 1.27 crash with no errors in logs

**Comment by Rayope** [comment]
I have a fix... in function Launch() of LauncherLogic.cs, line 945:

Replace this:
```
if (!AuthStateOverWrite)  
{  
    HelperClasses.Logger.Log("Letting Dragon work his magic");  
    await ROSCommunicationBackend.GenLaunchToken();  
}  
```

By this:
```
if (!AuthStateOverWrite)
{
  HelperClasses.Logger.Log("Letting Dragon work his magic");
  try
  {
      await ROSCommunicationBackend.GenLaunchToken();
  }
  catch(System.Net.Http.HttpRequestException exception)
  {
      HelperClasses.Logger.Log("Unable to connect to the server.");
      new Popups.Popup(Popups.Popup.PopupWindowTypes.PopupOk, "Unable to connect to the server.").ShowDialog();
      return;
  }
}
```

**Comment by TwosHusbandS** [comment]
Thanks again for the fix and code snippet and everything, as communicated via discord.

Closing this since it has been fixed in the 1.2.3.0 Release.
